### PR TITLE
Altered PdoConnection to allow nested transaction.

### LINKED ISSUE #4: Alter to allow for nested transactions.

### PR DESCRIPTION
**Pull request by JReissmueller**
Created/altered tests to account for this.

Fixes #4

### CODE REVIEW

File: src/PdoConnection.php [hunk 1]
```diff
@@ -58,6 +58,16 @@ class PdoConnection
...
 
     /**
+     * @var int Number of open transactions
+     */
+    private $transactionCount = 0;
+
+    /**
+     * @var bool Flag that tells whether a rollback is needed
+     */
+    private $rollbackNeeded = false;
```
Comment (clphillips) [inline src/PdoConnection.php:68]: Why have a flag that does that same thing as `$transactionCount === 0`? Also the naming is misleading. Just because a transaction has been processed doesn't mean it needs to be rolled back.

File: src/PdoConnection.php [hunk 2]
```diff
@@ -228,7 +238,11 @@ public function prepare($sql, $fetchMode = null)
      */
     public function begin()
     {
-        return $this->connect()->beginTransaction();
+        if (!$this->transactionCount++) {
+            return $this->connect()->beginTransaction();
+        }
+
+        return ($this->transactionCount >= 0);
```
Comment (clphillips) [inline src/PdoConnection.php]: superfluous parens.

File: src/PdoConnection.php [hunk 3]
```diff
@@ -238,7 +252,13 @@ public function begin()
      */
     public function rollBack()
     {
-        return $this->connect()->rollBack();
+        if (--$this->transactionCount === 0) {
+            $this->rollbackNeeded = false;
+            return $this->connect()->rollBack();
+        }
+
+        $this->rollbackNeeded = true;
```
Comment (clphillips) [inline src/PdoConnection.php:262]: Again, flag not needed, but since it's here, kinda weird to set something to true when it must already be true to reach this line.

File: src/PdoConnection.php [hunk 4]
```diff
@@ -248,7 +268,17 @@ public function rollBack()
...
     public function commit()
     {
-        return $this->connect()->commit();
+        if (--$this->transactionCount === 0) {
+            if (!$this->rollbackNeeded) {
+                return $this->connect()->commit();
+            }
+
+            $this->rollbackNeeded = false;
+            $this->connect()->rollback();
```
Comment (clphillips) [inline src/PdoConnection.php]: ??? Rolling back within a commit has got to be some kind of crossing-the-streams kind of bad.

File: src/PdoConnection.php [hunk 5]
```diff
@@ -248,7 +268,17 @@ public function rollBack()
      */
     public function commit()
     {
-        return $this->connect()->commit();
+        if (--$this->transactionCount === 0) {
+            if (!$this->rollbackNeeded) {
```
Comment (clphillips) [inline src/PdoConnection.php]: (╯°□°)╯︵ ┻━┻

File: tests/Unit/PdoConnectionTest.php [hunk 6]
```diff
@@ -140,8 +140,13 @@ public function testBegin()
      */
     public function testRollBack()
     {
+        // begin and rollback transaction
         $this->connection->setConnection($this->mockConnection('rollBack', true));
+        $this->connection->begin();
         $this->assertTrue($this->connection->rollBack());
+
+        // Since there is no open transaction, this should return false
+        $this->assertFalse($this->connection->rollBack());
```
Comment (clphillips) [inline tests/Unit/PdoConnectionTest.php:148]: comment is unnecessary. The assert explains what's expected

File: tests/Unit/PdoConnectionTest.php [hunk 7]
```diff
@@ -153,8 +158,60 @@ public function testRollBack()
      */
     public function testCommit()
     {
+        // begin and commit transaction
```
Comment (clphillips) [inline tests/Unit/PdoConnectionTest.php]: comment just makes this less readable, imo.

File: tests/Unit/PdoConnectionTest.php [hunk 8]
```diff
@@ -153,8 +158,60 @@ public function testRollBack()
...
         $this->connection->setConnection($this->mockConnection('commit', true));
+        $this->connection->begin();
         $this->assertTrue($this->connection->commit());
+
+        // Since there is no open transaction, this should return false
+        $this->assertFalse($this->connection->commit());
+    }
+
+    public function nestedTransactionData()
```
Comment (clphillips) [inline tests/Unit/PdoConnectionTest.php:173]: missing doc comments

File: src/PdoConnection.php [hunk 9]
```diff
@@ -238,7 +252,13 @@ public function begin()
      */
     public function rollBack()
     {
-        return $this->connect()->rollBack();
+        if (--$this->transactionCount === 0) {
```
Comment (clphillips) [inline src/PdoConnection.php:257]: have you considered what happens if you call `rollBack()` before calling `begin()`? :smiley:

File: src/PdoConnection.php [hunk 10]
```diff
@@ -248,7 +268,17 @@ public function rollBack()
      */
     public function commit()
     {
-        return $this->connect()->commit();
+        if (--$this->transactionCount === 0) {
```
Comment (clphillips) [inline src/PdoConnection.php:275]: Again, suppose this is called before `begin()`

File: src/PdoConnection.php [hunk 11]
```diff
@@ -248,7 +268,17 @@ public function rollBack()
...
+            if (!$this->rollbackNeeded) {
+                return $this->connect()->commit();
+            }
+
+            $this->rollbackNeeded = false;
+            $this->connect()->rollback();
+            return false;
+        }
+
+        return ($this->transactionCount >= 0 && !$this->rollbackNeeded);
```
Comment (clphillips) [inline src/PdoConnection.php]: `return` `return` `return` :sob: 

imo, the only time it makes sense to return something other than `false` is if you `return $this->connect()->commit()`.

File: tests/Unit/PdoConnectionTest.php [hunk 12]
```diff
@@ -153,8 +158,60 @@ public function testRollBack()
...
+     *
+     * @param array $actions
+     * @param bool $return
+     */
+    public function testNestedTransactions(array $actions, $return)
+    {
+        $this->connection->setConnection($this->mockConnection('beginTransaction', true));
+        foreach ($actions as $action) {
+            $this->assertTrue($this->connection->begin());
+        }
```
Comment (clphillips) [inline tests/Unit/PdoConnectionTest.php]: probably ought to add `'begin'` to the list of actions in the data provider.

File: tests/Unit/PdoConnectionTest.php [hunk 13]
```diff
@@ -153,8 +158,60 @@ public function testRollBack()
...
+
+        end($actions);
+        $end_index = key($actions);
+        reset($actions);
```
Comment (clphillips) [inline tests/Unit/PdoConnectionTest.php]: (╯°□°)╯︵ ┻━┻

```php
$end_index = count($actions) - 1;
```

File: tests/Unit/PdoConnectionTest.php [hunk 14]
```diff
@@ -153,8 +158,60 @@ public function testRollBack()
...
+        foreach ($actions as $index => $action) {
+            if ($index === $end_index) {
+                if ($return) {
+                    $this->connection->setConnection($this->mockConnection($action, true));
+                }
+                $this->assertEquals($return, $this->connection->{$action}());
+            } else {
+                $this->connection->{$action}();
+            }
+        }
```
Comment (clphillips) [inline tests/Unit/PdoConnectionTest.php:225]: could be written

```php
$end = count($actions) -1;
foreach ($actions as $index => $action) {
    if ($index === $end) {
        $this->connection->setConnection($this->mockConnection($action, $expected);
    }
    $actual = $this->connection->{$action}();
    $index === $end && $this->assertEquals($expected, $actual);
}
```

File: tests/Unit/PdoConnectionTest.php (hunk 13 above)
Comment (clphillips) [reply tests/Unit/PdoConnectionTest.php]: does PSR2 allow underscore in variable names? I thought they had to be `camelCase`...

**Review by clphillips — CHANGES_REQUESTED** [review]
I don't like this for a number of reasons:

1. It doesn't allow nested transactions. It simply silently ignores when `begin()` is called multiple times, then does nothing when `commit()` is called, unless you call it an equal number of times.
2. As mentioned above, calling `commit()` does nothing in some cases. That's unexpected behavior, and a sinkhole for bugs.
3. This does nothing to support real nested transactions on systems that support nested transactions, like innoDB. Nor can is solve the issue of implicit transactions (i.e. DROP TABLE) within a nested call set.
4. Nested transaction calls can easily be avoided using a service layer. Controller calls Service, Service calls begin/commit/rollback around the models to perform a particular action.

**Comment by JReissmueller** [comment]
So I could fix a lot of the things you commented on, but it looks like you disagree with this change as a concept.  It did seem that you were not quite understanding the intention for some of the code though so I'll try and clear that up.

You were totally correct in saying that this does not actually implement nested transactions, but is purely meant to ignore them so that one function that is currently using a transaction does not need to worry about calling another one that also uses a transaction.  To that end it is designed to, as you said, only begin on the first begin, and commit on the last commit.

The rollbackNeeded function will only be true when one of the "nested" transactions tries to rollback its changes,  In this case, we do not want to rollback immediately, because we would then end our transaction before reaching the end of the encompassing transaction.  Still, we do need to acknowledge that because the inner transaction failed, so should the outer.  So in the event that the outer succeeds and tries to commit, I decided to rollback and return false based on "rollbackNeeded".

In response to your conceptual objections:

1. You are totally correct.
2. You can tell when a commit failed by the return value.  Not sleek, but it is predictable.
3. As you stated in (1), this is not meant to truly implement nested transactions, only remove the error of calling a transaction within another.  As far as implicit commits, I'm not sure what that has to do with nested transactions in particular (unless you meant something different by "implicit transactions").
4. Not really sure what you mean here.  Are you saying to implement this behavior, just in a separate layer?

File: src/PdoConnection.php (hunk 11 above)
Comment (JReissmueller) [reply src/PdoConnection.php]: The return is meant to signal whether an error occurred.   In the case were we have not reached the outer transaction, and none of the inner transactions have tried to rollback, we return true for success.

File: src/PdoConnection.php (hunk 4 above)
Comment (JReissmueller) [reply src/PdoConnection.php]: Yea, it feels a bit weird, but if one of the inner transactions fails, then we must not commit the outer transaction.  We return false to indicate that this has happened.

File: src/PdoConnection.php (hunk 10 above)
Comment (JReissmueller) [reply src/PdoConnection.php:275]: I could try to make it more robust for handling cases like these, but that seems like a pretty clear dev error.

File: src/PdoConnection.php (hunk 3 above)
Comment (JReissmueller) [reply src/PdoConnection.php:262]: The flag is false by default and is only set to true when a rollback is called on an inner transaction.  It doesn't depend on the current value of rollbackNeeded at all.

File: src/PdoConnection.php (hunk 9 above)
Comment (JReissmueller) [reply src/PdoConnection.php:257]: Yep, relatively bad ju-ju-magumbo.  Counter will go negative and rollback needed will be set to false.  If a corresponding number of begins are called afterward, that the counter will return to 0 and the only consequence will be that rollbackNeeded will be set to true.  Which is actually pretty bad.  Like I mentioned in my other comment, this seems like a pretty obvious dev error.

File: src/PdoConnection.php (hunk 1 above)
Comment (JReissmueller) [reply src/PdoConnection.php:68]: $transactionCount represents how many transactions are currently open.  $rollbackNeeded is only true when a rollback is called on a transaction within another transaction, which is very different from $transactionCount === 0.  I means that if the outer transaction tries to commit, we need to rollback instead.

Comment (clphillips) [reply src/PdoConnection.php:68]: Performing a rollback within a commit is not a good design choice, imo. If you can't commit, or you can't rollback you probably need to throw an exception because that's definitely programmer error.

Also, if this is just a queue system, why not use a queue?

**Comment by clphillips** [comment]
I presume the reason for this PR is to silently ignore when `begin()` is called after already having been called, and `commit()` if more than 1 `begin()` has queued. I'm guessing the logic is something like:

```php
class Model {
    public function saveThisThing() {
        $db->begin();
        // stuff
        $this->saveThatThing();
        $db->commit();
    }

    public function saveThatThing() {
        $db->begin();
        // stuff
        $db->commit();
    }
```

To solve this on a service layer you'd simply execute a single transaction per service call. It would look something like:

```php
class ThisThingService {
    public function createThisThing(Model $model) {
        $db->begin();
        $model->saveThisThing();
        $model->saveThatThing();
        $db->commit();
    }
}

class Model {
    public function saveThisThing() {
        // stuff
    }

    public function saveThatThing() {
        // stuff
    }
}
```

If you needed to `saveThisThing()` without `saveThatThing()` somewhere else, you'd just add another method to your `ThisThingService`. Much more flexible.

Basically what I'm getting at is that nested transactions are a lot like buffered queries. If you need them, you're doing something wrong.
